### PR TITLE
308 Topbar lost when moving between monitors

### PR DESCRIPTION
**Pull request by elshimone**
See #308  Seems to be a conflict with dash to dock.  Fix is to only move topbar
when topbar_follow_focus is toggled off.

### CODE REVIEW

File: topbar.js [hunk 1]
```diff
@@ -622,8 +625,15 @@ function setWorkspaceName (name) {
 }
 
 function setMonitor(monitor) {
-    if (!prefs.topbar_follow_focus)
```
Comment (hedning) [inline topbar.js:625]: When `topbar_follow_focus` is true we need to move the top bar to `monitor`. 

 `panelMonitor` also needs to be set as there's other code which depend on knowing which monitor the top bar is on (not necessary each time, but it doesn't hurt).

File: topbar.js [hunk 2]
```diff
@@ -539,11 +539,14 @@ function enable () {
 
     signals.connect(Settings.settings, 'changed::topbar-follow-focus', (settings, key) => {
         let monitors = Tiling.spaces.monitors;
-        let to = setMonitor(Main.layoutManager.focusMonitor);
-        let space = monitors.get(to);
-        updateWorkspaceIndicator(space.workspace.index());
-        for (let [workspace, space] of Tiling.spaces) {
-            space.layout();
+        if (!settings.prefs.topbar_follow_focus) {
```
Comment (hedning) [inline topbar.js:542]: We want to do everything here when `topbar_follow_focus` is true too. (eg. layouts should be run to fix up window heights).

So we only want to move the top bar to the primary monitor in this if block.

I guess it makes sense to generalize `moveTopBarToPrimary()` into `moveTopBarTo(monitor)`, that way `setMonitor` cas use it too when top bar should follow focus.

**Comment by elshimone** [comment]
Ok I understand what the topbar_follow_focus setting does, and it appears to work with dash to dock disabled (i.e. the top bar moves to the focussed monitor).  But if I turn on topbar_follow_focus and dash to dock then I get the original issue :(  At least I can turn off the setting for my own use, but I'd like to fix this properly.

**Comment by hedning** [comment]
Yeah, better support for dash-to-panel would be nice, but I'm not really sure if it's reasonable to fix it when top bar follow focus unfortunately, though I might be wrong here.

As I understand it dash-panel have a lot of options, one of which puts a panel on every monitor. In that case paperwm shouldn't be moving the top bar around at all, but without some ad-hoc dash-to-panel fix we can't necessarily know what to do.

I guess we could turn `topbar-follow-focus` off if we detect dash-to-panel, that's at least a minimal hack. 

(There's also an issue with how we handle the workarea, which makes having the panel on the bottom broken, I think the underlying issue might be fixed in newer gnome shells though making it possible to rely entirely on the reported workarea).

File: topbar.js [hunk 3]
```diff
@@ -622,8 +623,14 @@ function setWorkspaceName (name) {
 }
 
 function setMonitor(monitor) {
-    if (!prefs.topbar_follow_focus)
-        monitor = Main.layoutManager.primaryMonitor;
```
Comment (hedning) [inline topbar.js:626]: This needs to happen to. `setMonitor` is called when changing workspaces, and without it `panelMonitor` will be incorrect.

File: topbar.js [hunk 4]
```diff
@@ -539,13 +539,14 @@ function enable () {
 
     signals.connect(Settings.settings, 'changed::topbar-follow-focus', (settings, key) => {
         let monitors = Tiling.spaces.monitors;
-        let to = setMonitor(Main.layoutManager.focusMonitor);
-        let space = monitors.get(to);
-        updateWorkspaceIndicator(space.workspace.index());
```
Comment (hedning) [inline topbar.js:544]: We still want these lines too. `setMonitor` updates `panelMonitor` which the tiling code uses to know which monitor the panel is located. `updateWorkspaceIndicator` refreshes the workspace name in the top bar.

File: topbar.js [hunk 5]
```diff
@@ -539,13 +539,15 @@ function enable () {
 
     signals.connect(Settings.settings, 'changed::topbar-follow-focus', (settings, key) => {
         let monitors = Tiling.spaces.monitors;
+        if (!settings.prefs.topbar_follow_focus) {
+            moveTopBarTo(Main.layoutManager.prim);
```
Comment (hedning) [inline topbar.js:543]:
```suggestion
            moveTopBarTo(Main.layoutManager.primaryMonitor);
```

**Comment by hedning** [comment]
Cheers :) I'll try to test it out quickly and then merge.

**Comment by hedning** [comment]
Right, finally got a chance to take a look at this.

Seems like it mostly works as intended, but due to the way we handle monitor workarea there's some wonkyness left to be sorted out (the window is forced down on the second monitor). We should really just trust what the workarea method, I'll need to look closer why we made things as they are (my recollection is that fullscreen in older gnome shell versions caused the workarea to change which could cause trouble).

Though, what I mentioned above isn't a regression, so I think it's fine to merge this as is, and fix the remaining problems later.

**Comment by hedning** [comment]
Cheers :)

Sorry it took so long to merge.

**Pull request merged**